design/ceph-mon-pv.md

```diff
@@ -0,0 +1,97 @@
+# Ceph monitor PV storage
+
+**Target version**: Rook 1.1
+
+## Overview
+
+Currently all of the storage for Ceph monitors (data, logs, etc..) is provided
+using HostPath volume mounts. Supporting PV storage for Ceph monitors in
+environments with dynamically provisioned volumes (AWS, GCE, etc...) will allow
+monitors to migrate without requiring the monitor state to be rebuilt, and
+avoids the operational complexity of dealing with HostPath storage.
+
+The general approach taken in this design document is to augment the CRD with a
+persistent volume claim template describing the storage requirements of
+monitors. This template is used by Rook to dynamically create a volume claim for
+each monitor.
+
+## Monitor storage specification
+
+The monitor specification in the CRD is updated to include a persistent volume
+claim template that is used to generate PVCs for monitor database storage.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+The `VolumeClaimTemplate` is used by Rook to create PVCs for monitor storage.
+The current set of template fields used by Rook when creating PVCs are
+`StorageClassName` and `Resources`. Rook follows the standard convention of
+using the default storage class when one is not specified in a volume claim
+template. If the storage resource requirements are not specified in the claim
+template, then Rook will use a default value. This is possible because unlike
+the storage requirements of OSDs (xf: StorageClassDeviceSets), reasonable
+defaults (e.g. 5-10 GB) exist for monitor daemon storage needs.
+
+*Logs and crash data*. The current implementation continues the use of a
+HostPath volume based on `dataDirHostPath` for storing daemon log and crash
+data. This is a temporary exception that will be resolved as we converge on an
+approach that works for all Ceph daemon types.
+
+Finally, the entire volume claim template may be left unspecified in the CRD
+in which case the existing HostPath mechanism is used for all monitor storage.
+
+## Upgrades and CRD changes
+
+When a new monitor is created it uses the _current_ storage specification found
+in the CRD. Once a monitor has been created, it's backing storage is not
+changed. This makes upgrades particularly simple because existing monitors
+continue to use the same storage.
+
+Once a volume claim template is defined in the CRD new monitors will be created
+with PVC storage. In order to remove old monitors based on HostPath storage
+first define a volume claim template in the CRD and then fail over each monitor.
+
+## Clean up
+
+Like `StatefulSet`s removal of an monitor deployment does not automatically
```

Comment:
"Like `StatefulSets` **[<- nit: I think including the plural in the quote is fine since k8s defines plurals as well as the singular]**, removal of **a [<- typo]** monitor ..."